### DISCUSSION TITLE
How to map TSToggle to `J` and have the fallback be the original `J`?

### DISCUSSION THREAD

**Discussion by lfilho** [Wansmer/treesj#152] [Q&A] [answered]
Hi!
My use case: I would like to use `J` as the keymap for toggling. But, if the plugin fails to join/split something, I'd like it to just execute vim's default `J` (just joining two lines regardless of its semantics, i.e. just removing the linebreak).

Is that possible and if so how to do it? Thanks!

**Comment by Wansmer** [comment]
Hi! You can to do this with the `fallback` option and pass as fallback `vim.cmd.join()` or `vim.api.nvim_feedkeys('J', 'n', true)`.
How to create fallback:
1. https://github.com/Wansmer/treesj/discussions/126
2. https://github.com/Wansmer/treesj/issues/143#issuecomment-1891105670

**Comment by lfilho** [comment]
Hi @Wansmer thanks for your answer. I did see those before posting, but I still couldn't figure out how to apply to my case. Different from those, I don't want that to a particular language or conditional to Treesitter existing or not. As a matter of fact, what i was trying to accomplish is on a regular language with treesitter working in it (javascript).

What I would like is: 
1. invoke TSToggle as usual on my javascript file
2. It will try to do its thing
3. If it fails (no action from TSToggle), then call `vim.cmd.join()`

Hope that's clearer. Would you be able to guide me how to accomplish that? Thank you

**Reply by Wansmer** [reply]
The solution will be the same as [here](https://github.com/Wansmer/treesj/discussions/126). You just have to loop through each node (not just `comment` as in the example) in each language and add a fallback function with `vim.cmd.join()`.

Second solution, add a function to the error handler (`opts.on_error`) and call `join` in it. You can see the possible error texts in the file `lua/treesj/notify.lua`. However, this is a less predictable solution than the above option.

**Reply by lfilho** [reply]
I see, thanks. I'm not too familiar with lua nor the inner workings of your script.

Is this what it would look like?

```lua
    "Wansmer/treesj",
    keys = { { "J", "<cmd>TSJToggle<CR>", desc = "Split/Join" } },
    dependencies = { "nvim-treesitter/nvim-treesitter" },
    config = function()
      local langs = require("treesj.langs").presets

      -- Add fallback to all language
      for _, nodes in pairs(langs) do
        -- print nodes:

        for _, node in ipairs(nodes) do
          node.fallback = function()
            vim.cmd.join()
          end
        end
      end

      require("treesj").setup()
    end,
  },
```

It unfortunately didn't work. See screenshot below: i was just trying to get line 330 up (regular `J`). But TreeSJ just showed that notification, without actually calling `vim.cmd.join()`. Another weird things is the notification itself: it mentions containing a comment but as you can see in the screenshot, there isn't a comment in the code?!

![CleanShot 2024-05-17 at 10 45 39@2x](https://github.com/Wansmer/treesj/assets/282756/9f9e710b-2f8d-4ddf-97a5-836ec26f8fca)

Any clues? Thanks again

**Reply by Wansmer** [reply, marked as answer]
There's nothing wrong with your code. I guess I was wrong to advise you on this, sorry. Do this:

```lua
require("treesj").setup({
  on_error = function()
    vim.cmd.join()
  end,
})

vim.keymap.set("n", "J", require("treesj").join)
```

This will do what you want in all cases where `treesj` gets an error that for some reason it can't do `join` for a node.


https://github.com/Wansmer/treesj/assets/46977173/284943d7-7842-4d85-9b27-9ecf1150791f

> it mentions containing a comment but as you can see in the screenshot, there isn't a comment in the code?!

Read [how the plugin works](https://github.com/Wansmer/treesj?tab=readme-ov-file#how-plugin-works).

The nearest matching node has a comment

**Reply by lfilho** [reply]
Ah that seems to work!! Thank you!!! <3
